account_mtd/models/mtd_hello_world.py

```diff
@@ -0,0 +1,406 @@
+                    "Response Received: \n{}".format(response_token['message'])
+            )
+            self._logger.info(
+                "(Step 2) exchange authorisation code - log:- {}".format(construct_text)
+            )
+            action = self.env.ref('account_mtd.action_mtd_hello_world')
+            menu_id = self.env.ref('account_mtd.submenu_mtd_hello_world')
+            self._logger.info(
+                "(Step 2) exchange authorisation code - redirect URI :- " +
+                "/web#id={}&view_type=form&model=mtd.hello_world".format(record_id) +
+                "&menu_id=72&action={}".format(menu_id.id, action.id)
+            )
+            return werkzeug.utils.redirect(
+                '/web#id={}&view_type=form&model=mtd.hello_world&'.format(record_id) +
+                'menu_id=72&action={}'.format(menu_id.id, action.id)
+            )
+
+    def refresh_user_authorisation(self, token_record=None):
+        self._logger.info("(Step 4) refresh_user_authorisation - token_record:- {}".format(token_record))
+        api_token = self.env['mtd.api_tokens'].search([('id', '=', token_record.id)])
+        hmrc_authorisation_url = "{}{}".format(self.hmrc_configuration.hmrc_url, '/oauth/token')
+        self._logger.info(
+            "(Step 4) refresh_user_authorisation - hmrc authorisation url:- {}".format(hmrc_authorisation_url)
+        )
+
+        data_user_info = {
+            'client_secret': self.hmrc_configuration.client_secret,
+            'client_id': self.hmrc_configuration.client_id,
+            'grant_type': 'refresh_token',
+            'refresh_token': api_token.refresh_token
+            }
+        headers = {
+            'Content-Type': 'application/json',
+            }
+        self._logger.info(
+            "(Step 4) refresh_user_authorisation - data to send in request:- {}".format(data_user_info)
+        )
+        self._logger.info(
+            "(Step 4) refresh_user_authorisation - headers to send in request:- {}".format(headers)
+        )
+        req = requests.post(hmrc_authorisation_url, data=json.dumps(data_user_info), headers=headers)
+        response_token = json.loads(req.text)
+        self._logger.info(
+            "(Step 4) refresh_user_authorisation - received response " +
+            "of the request:- {}, and its text:- {}".format(req, response_token)
+        )
+        if req.ok:
+            api_token.access_token = response_token['access_token']
+            api_token.refresh_token = response_token['refresh_token']
+            api_token.expires_in = json.dumps(response_token['expires_in'])
+            version = self._json_command('version')
+            return version
+        elif req.status_code == 400 and response_token['message'] == "Bad Request":
+            self._logger.info(
+                "(Step 4) refresh_user_authorisation - error 400, Need to obtain new access code and refresh token"
+            )
+            self.get_user_authorisation()
```

Comment:
Not returning anything.  This will upset any XML-RPC callers.

Reply:
Does get_user_authorisation() obtain the new access code and refresh token?

If yes, consider this log message:

```
(Step 4) refresh_user_authorisation - error 400, obtaining new access code and refresh token...
```